iree/compiler/Dialect/HAL/IR/HALTypes.cpp

```diff
@@ -297,16 +297,19 @@ enum class NumericalType : uint32_t {
   kIntegerUnsigned = 0x02,
   // TODO(benvanik): specialize with semantics from APFloat.
   kFloatIEEE = 0x03,
+  kFloatBrain = 0x04,
 };
 constexpr inline int32_t makeElementTypeValue(NumericalType numericalType,
                                               int32_t bitCount) {
   return (static_cast<uint32_t>(numericalType) << 24) | bitCount;
 }
 }  // namespace
+
 llvm::Optional<int32_t> getElementTypeValue(Type type) {
   if (auto intType = type.dyn_cast_or_null<IntegerType>()) {
-    // TODO(benvanik): add signed/unsigned check when landed in MLIR.
-    return makeElementTypeValue(NumericalType::kIntegerSigned,
+    return makeElementTypeValue(intType.isUnsigned()
```

Comment:
Seems like this should also take into account signless? Are we assuming signless => signed. I think after we get out of MHLO we're generally not dealing with signfull integers at all as the semantics have been transferred to the op. But maybe this method is dealing with ABI boundaries or printing, where we would still have them? I think probably worth a doc comment of what exactly the use case is though

Reply:
These are only used for printing text or reflection so it's best to match whatever the user originally had specified. If we start removing the information earlier in the pipeline we'll need to find a way to push it through to preserve this.

There's no such thing as a signless integer in C so we have to pick one - do you think signless should print/parse as unsigned or signed? (you know more about this than me :)

Reply:
actually, I think maybe I could just plumb through signless - we should at least figure out how we want to handle the parsing/printing though. I'll update the PR with signless.

Reply:
nevermind - there's no direct mapping of signless to just about anything so it just adds a bunch of additional footguns for handling the data types at the edges.

Reply:
Added comments about signless and cleaned things up a bit - still will need more end-to-end work but this is at least progress. If you think signless should be parsed/printed as unsigned lmk and I can flip that around.

Reply:
flip flopping; this is all nasty ;( can't round trip strings of the data types without supporting signless everywhere or not supporting it anywhere - going to fiddle with it

Reply:
ok this may work now and fixes the inconsistency that existed (where `u16` instead of `ui16` for unsigned). We should be able to represent all 3 types now but there's likely still work to be done on the ABI to ensure we are reflecting things correctly.